0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md

```diff
@@ -1,4 +1,263 @@
+Now compile and load it and watch red LED blinking. 
+
+---
+### BREATH (in C) ###
+Now we have blinking led, but I find it a little bit rough. Let’s make it glow. We will need PWM. According to babel schematic one of the LEDS is connected to channel4 of timer3.
+In order to enable timer3 pwm generation ChibiOS must be configured a little bit. 
+
+
+Find macro ```HAL_USE_PWM``` in ```halconf.h``` and make it `TRUE`
+
+Find macro `STM32_PWM_USE_TIM3` in ```mcuconf.h``` and make it `TRUE`
+
+Now add breath_thread to you code: 
+```c
+static PWMConfig pwmcfg =	//initialization structure for ChibiOS PWM hal 
+{
+    10000,        //10KHz PWM clock frequency.   
+    255,          //255 ticks is pwm resolution
+    NULL,
+    {
+        {PWM_OUTPUT_DISABLED, NULL},
+        {PWM_OUTPUT_DISABLED, NULL},
+        {PWM_OUTPUT_DISABLED, NULL},
+        {PWM_OUTPUT_ACTIVE_HIGH, NULL}
+    },
+    /* HW dependent part.*/
+    0,
+    0
+};
+
+static THD_WORKING_AREA(waThread2, 255);
+static THD_FUNCTION(breath_thread, arg) 
+{
+    (void)arg;
+    uint8_t brightness = 0;
+    chRegSetThreadName("breath");
+    palSetPadMode(GPIOB, 1, PAL_MODE_ALTERNATE(2));
+    pwmStart(&PWMD3, &pwmcfg);  			//Timer activation in PWM mode
+    pwmEnableChannelI(&PWMD3, 3, 128);	//Brightness setting
+    while (true) 
+    {   
+        while (brightness <= 250)
+        {
+            pwmEnableChannelI(&PWMD3, 3, brightness++);
+            chThdSleepMilliseconds(3);
+        }
+        while (brightness >= 5)
+        {
+            pwmEnableChannelI(&PWMD3, 3, brightness--);
+            chThdSleepMilliseconds(3);
+        }
+    }
+}
+```
+And change your ```main()``` to start this thread: 
+```c
+int main()
+{
+    chThdCreateStatic(waThread1, sizeof(waThread1), NORMALPRIO+1, blink_thread,  NULL);
+    chThdCreateStatic(waThread2, sizeof(waThread2), NORMALPRIO+1, breath_thread, NULL);
+    while(1)
+    {
+    }
+}
+```    
+
+Load it and watch green led blinking ON and OFF and red led smoothly chaging its brightness. Isn’t it fancy?
+
+---
+
+### BLINK and BREATH in C++ ###
+Now when we can blink a LED differently lets make it in a more civilized way - in C++
+Although Chibios is written in C it offers rich C++ capabilities and we can rewrite out blinking threads like this:
+
+```
+class BlinkerThread : public chibios_rt::BaseStaticThread<128>
+{
+    void main() override
+    {
+        setName("blinker");
+        palSetPadMode(GPIOE, 8, PAL_MODE_OUTPUT_PUSHPULL);
+        while (true)
+        {
+            palSetPad(GPIOE,8);
+            chThdSleepMilliseconds(100);
+            palClearPad(GPIOE, 8);
+            chThdSleepMilliseconds(100);
+        }
+    }
+
+public:
+    virtual ~BlinkerThread() { }
+} blinker_thread_;
+```
+
+and
+
+```
+static const PWMConfig pwmcfg = {
+  10000,                                    //10KHz PWM clock frequency.   
+  255,                                      //255 ticks is pwm resolution                                    
```

Comment:
Missing space at the beginning of the comments. I admit our style guide does not say anything about this, this is my fault; I will update it right now.